### DISCUSSION TITLE
How to disable line weights when saving a page to image?

### DISCUSSION THREAD

**Discussion by cpwan** [pymupdf/PyMuPDF#1464] [Looking for help] [answered]
I am currently working on pages with vector graphics. I want to save the page as an image but I encounter issues with 'line weights'. 

## What I expected: (with line weights disabled)
![image](https://user-images.githubusercontent.com/22132868/146327734-f01a4a9f-d1da-488a-a398-9fae08f7af24.png)
## What I got: (with line weights enabled)
![image](https://user-images.githubusercontent.com/22132868/146327757-71c7b25d-9854-4e51-b9d4-95f0d58b6bbb.png)

In adobe pdf, I can disable 'line weights' when ( and only when) viewing the pdf, following this [link](https://www.engipedia.com/show-thin-lines-in-pdf/) .

I am using the following code to save the page as image:
```python
doc = fitz.open(pdffile)
page = doc.loadPage(page_id)
pix = page.getPixmap(annots=False, matrix=fitz.Matrix(scale, scale)) 
pix.writePNG(out_name)
```

How can I disable the 'line weights' when saving it as image?

Here is a minimal example. The difference becomes more prominent when zooming in.
> [minimal example.pdf](https://github.com/pymupdf/PyMuPDF/files/7725553/minimal.example.pdf)

**Comment by JorjMcKie** [comment]
May be a problem of your pymupdf version ... at least you are using the deprecated camelCase names.
I did this:
```python
>>> for dpi in (100, 200, 300):
	pix = page.get_pixmap(dpi=dpi)
	pix.save("%i.png" % dpi)
```
... and everything looks perfect!
100 dpi:
![grafik](https://user-images.githubusercontent.com/8290722/146345067-94ea55aa-8378-48f5-9322-4a6b0c56ad95.png)
200 dpi:
![grafik](https://user-images.githubusercontent.com/8290722/146345170-e7e6b970-e9ae-453c-a4bd-14d07ad9fe59.png)
300 dpi:
![grafik](https://user-images.githubusercontent.com/8290722/146345278-222bce89-e35c-488e-b913-ded470a0f79a.png)

And to be sure, here one with 400 dpi:
![grafik](https://user-images.githubusercontent.com/8290722/146345906-c62b2d34-a073-4ebf-9d3b-f327f7f35e2d.png)


pymupdf version: 1.19.3

**Comment by cpwan** [comment]
I expect the following, with the 'line weight' disabled. This is viewed with adobe pdf.
![image](https://user-images.githubusercontent.com/22132868/146347201-2a26779b-80ca-4a27-8de7-46be26d4a3af.png)

**Reply by cpwan** [reply]
https://user-images.githubusercontent.com/22132868/146349710-76994739-ed90-48de-8f09-3543d9a35a21.mp4

**Comment by JorjMcKie** [comment]
So you want an option to maintain a line width of 1 even if the image is generated with 400% resolution?
No that is not possible.
Pixmap generation scales everything evenly.

**Comment by JorjMcKie** [comment]
But your first image shown with that CL600 text should not happen either.
Can you let me have the corresponding file?

**Reply by cpwan** [reply]
Sure, please have a look: [cl600 demo.pdf](https://github.com/pymupdf/PyMuPDF/files/7731440/cl600.demo.pdf)

**Comment by cpwan** [comment]
My current workaround is to 
1. save the page as svg
2. modify the stroke-width to a small value like .01 
3. print the svg to pdf
4. use the above code to save it as an image

```python
import re
doc = fitz.open(pdffile)
page = doc.loadPage(page_id)
svg = page.get_svg_image(matrix=fitz.Identity)
svg = re.sub(r'stroke-width=".[0-9][0-9]"', 'stroke-width=".01"', svg)
print(svg, file=open("dummy.svg", "w"))
pix_doc = fitz.open("dummy.svg")
pix = pix_doc.load_page(0).get_pixmap(
     annots=False, matrix=fitz.Matrix(scale, scale)
)
pix.save("thin_lines.png")
```

The problem becomes how to change the line width of vector graphics in pdf.

**Comment by JorjMcKie** [comment, marked as answer]
This snippet also works ... at least with your example PDF:
```python
>>> doc=fitz.open("cl600.demo.pdf")
>>> page=doc[0]
>>> page.clean_contents() # enforce standard painting syntax
>>> cont_xref = page.get_contents()[0] # we only have 1 of those now
>>> # read an modify the paint commands
>>> lines = doc.xref_stream(cont_xref).splitlines()
>>> # walk through lines and replace each width value
>>> for i in range(len(lines)):
	if lines[i].endswith(b" w"):  # command to set line width!
		lines[i] = b"0.02 w"

		
>>> # write back modified paint commands
>>> doc.update_stream(cont_xref, b"\n".join(lines))
>>> # now either make a pixmap of modified page ...
>>> pix=page.get_pixmap(dpi=300)
>>> pix.save("300.png")
>>> # ... or save a modified PDF
>>> doc.save("thin-lines.pdf")
>>> 
```

**Reply by cpwan** [reply]
Thanks!
